### PR TITLE
Basic task completed

### PR DESCRIPTION
**Pull request by ljden**

### CODE REVIEW

File: lab04_example/package.xml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="mtrn4230@todo.todo">mtrn4230</maintainer>
+
+
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
```
Comment (markwhittyunsw) [inline lab04_example/package.xml]: Suggest use of MIT license

File: lab04_example/package.xml [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>lab04_example</name>
+  <version>0.0.0</version>
+  <description>The lab04_example package</description>
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="mtrn4230@todo.todo">mtrn4230</maintainer>
```
Comment (markwhittyunsw) [inline lab04_example/package.xml]: Recommend students update this to their own contact details.

Comment (markwhittyunsw) [inline lab04_example/package.xml]: Recommend students update this to their own contact details.

File: lab04_example/package.xml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+
+  <!-- Author tags are optional, multiple are allowed, one per tag -->
+  <!-- Authors do not have to be maintainers, but could be -->
+  <!-- Example: -->
+  <!-- <author email="jane.doe@example.com">Jane Doe</author> -->
+
+
+  <!-- The *depend tags are used to specify dependencies -->
+  <!-- Dependencies can be catkin packages or system dependencies -->
```
Comment (markwhittyunsw) [inline lab04_example/package.xml]: See conceptual overview: http://wiki.ros.org/catkin/conceptual_overview

File: lab04_example/CMakeLists.txt [hunk 4]
```diff
@@ -1,206 +1,36 @@
...
 
-## Find catkin macros and libraries
-## if COMPONENTS list like find_package(catkin REQUIRED COMPONENTS xyz)
-## is used, also find other catkin packages
 find_package(catkin REQUIRED COMPONENTS
-  geometry_msgs
   roscpp
   std_msgs
+  geometry_msgs
+  genmsg
```
Comment (markwhittyunsw) [inline lab04_example/CMakeLists.txt:12]: Most lines in this file need further comments (unless added in later commits)

File: lab04_example/package.xml (hunk 1 above)
Comment (ljden) [reply lab04_example/package.xml]: You seem to be commenting on old versions of the files?

File: lab04_example/include/lab04_example/VelocityController.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace Lab04 {
+    /* Inspiration from:
+     *  https://github.com/wsnewman/ros_class/blob/master/example_ros_class/src/example_ros_class.h
+     *  https://roboticsbackend.com/roscpp-timer-with-ros-publish-data-at-a-fixed-rate/
+     */
+    class VelocityController {
+        public:
+            VelocityController(ros::NodeHandle nh) : nh_(nh) {
+                key_control_sub_ = nh_.subscribe("KeyControl", 10,
```
Comment (markwhittyunsw) [inline lab04_example/include/lab04_example/VelocityController.h:21]: Recommend some commentary around variable style here, or links to resources discussing style.

File: lab04_example/include/lab04_example/VelocityController.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                    case(KeyControl::A_KEY):
+                    case(KeyControl::S_KEY):
+                    case(KeyControl::D_KEY):
+                    case(KeyControl::Q_KEY):
+                    case(KeyControl::E_KEY):
+                    case(KeyControl::SPACE_KEY):
+                        current_cmd_ = static_cast<KeyControl::key_press>(msg->data);
+                        break;
+                    default:
+                        ROS_ERROR("Recieved invalid command: %d", msg->data);
```
Comment (markwhittyunsw) [inline lab04_example/include/lab04_example/VelocityController.h]: typo: received

File: lab04_example/include/lab04_example/VelocityController.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        private:
+            ros::NodeHandle nh_;
+
+            ros::Subscriber key_control_sub_;
+            ros::Publisher 	twist_pub_;
+            ros::Timer cmd_vel_timer_;
+            ros::ServiceClient client_;
+
+            KeyControl::key_press current_cmd_{KeyControl::NULL_KEY};
```
Comment (markwhittyunsw) [inline lab04_example/include/lab04_example/VelocityController.h:36]: Comment on what this is doing, C++ familiarity is low

File: lab04_example/include/lab04_example/VelocityController.h [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+#include <lab04_example/key_control.h>
+#include <lab04_example/KeyToCmdVel.h>
+
+namespace Lab04 {
+    /* Inspiration from:
+     *  https://github.com/wsnewman/ros_class/blob/master/example_ros_class/src/example_ros_class.h
+     *  https://roboticsbackend.com/roscpp-timer-with-ros-publish-data-at-a-fixed-rate/
+     */
+    class VelocityController {
+        public:
```
Comment (markwhittyunsw) [inline lab04_example/include/lab04_example/VelocityController.h:19]: Needs comments about what this class will do, how it relates to the lab

File: lab04_example/include/lab04_example/VelocityController.h [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                    case(KeyControl::E_KEY):
+                    case(KeyControl::SPACE_KEY):
+                        current_cmd_ = static_cast<KeyControl::key_press>(msg->data);
+                        break;
+                    default:
+                        ROS_ERROR("Recieved invalid command: %d", msg->data);
+                }
+            }
+
+            auto cmdVelCallback_() -> void {
```
Comment (markwhittyunsw) [inline lab04_example/include/lab04_example/VelocityController.h:59]: Comments on each method need, when is it called, why is it used?

File: lab04_example/include/lab04_example/key_control.h [hunk 10]
```diff
@@ -0,0 +1,54 @@
+/* Author: Luke Dennis l.dennis@unsw.edu.au */
+
+#include <ros/ros.h>
+#include <termios.h>
+
+namespace KeyControl {
+    /* https://github.com/sdipendra/ros-projects/blob/master/src/keyboard_non_blocking_input/src/keyboard_non_blocking_input_node.cpp*/
+    char getch() {
+        fd_set set;
```
Comment (markwhittyunsw) [inline lab04_example/include/lab04_example/key_control.h:9]: what is fd_set?

File: lab04_example/src/key_input.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        if(key != KeyControl::NULL_KEY) {
+            /* send every non-null key */
+            no_key_count = 0;
+
+            key_msg.data = key;
+            key_control_pub.publish(key_msg);
+        } else {
+            ROS_INFO_DELAYED_THROTTLE(10,"Press Space to quit");
```
Comment (markwhittyunsw) [inline lab04_example/src/key_input.cpp:95]: Comment this call.

File: lab04_example/src/key_to_cmd_vel_server.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
+/* http://wiki.ros.org/ROS/Tutorials/WritingServiceClient%28c%2B%2B%29#roscpp_tutorials.2FTutorials.2FWritingServiceClient.Writing_a_Service_Node */
+
+/* Author: Luke Dennis l.dennis@unsw.edu.au */
+
+#include <ros/ros.h>
+#include <lab04_example/KeyToCmdVel.h>
+#include <lab04_example/key_control.h>
+
+
+class KeyCmdSrv {
```
Comment (markwhittyunsw) [inline lab04_example/src/key_to_cmd_vel_server.cpp:14]: Document how this class relates to all other classes

File: lab04_example/src/key_to_cmd_vel_server.cpp [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+        geometry_msgs::Twist last_vel_;
+
+        auto convert_to_vel_(lab04_example::KeyToCmdVel::Request & req,
+                lab04_example::KeyToCmdVel::Response & res) -> bool {
+
+            ROS_INFO("(%lf, %lf) (%lf, %lf)", last_vel_.linear.x, last_vel_.linear.y,
+                    last_vel_.angular.x, last_vel_.angular.y);
+
+            /* VERY basic service, this can be extended to do basic DR based on
```
Comment (markwhittyunsw) [inline lab04_example/src/key_to_cmd_vel_server.cpp]: dead reckoning

File: lab04_example/src/velocity_command_node.cpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* Author: Luke Dennis l.dennis@unsw.edu.au */
+
```
Comment (markwhittyunsw) [inline lab04_example/src/velocity_command_node.cpp:29]: Add brief comment on what this node does

File: lab04_example/srv/KeyToCmdVel.srv [hunk 15]
```diff
@@ -0,0 +1,4 @@
+# Author: Luke Dennis l.dennis@unsw.edu.au
```
Comment (markwhittyunsw) [inline lab04_example/srv/KeyToCmdVel.srv:1]: Add link to where info on services can be found

**Review by markwhittyunsw — CHANGES_REQUESTED** [review]
Recommendations mostly for comments.

File: lab04_example/package.xml (hunk 1 above)
Comment (markwhittyunsw) [reply lab04_example/package.xml]: yes, sorry, I started at the beginning, will work backwards now

File: lab04_example/CMakeLists.txt (hunk 4 above)
Comment (ljden) [reply lab04_example/CMakeLists.txt:12]: These are auto generated and walked through in the PublisherSubscriber tutorial on ROS http://wiki.ros.org/ROS/Tutorials/WritingPublisherSubscriber%28c%2B%2B%29#roscpp_tutorials.2FTutorials.2FWritingPublisherSubscriber.Building_your_nodes

File: lab04_example/include/lab04_example/VelocityController.h (hunk 5 above)
Comment (ljden) [reply lab04_example/include/lab04_example/VelocityController.h:21]: I think this is outside the scope but here are some (dense) resources:
https://google.github.io/styleguide/cppguide.html
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines

File: lab04_example/src/key_input.cpp (hunk 11 above)
Comment (ljden) [reply lab04_example/src/key_input.cpp:95]: the publish?

File: lab04_example/include/lab04_example/VelocityController.h (hunk 8 above)
Comment (ljden) [reply lab04_example/include/lab04_example/VelocityController.h:19]: 4f5e9f7b18aad97f02a4c2dc5ecaa2cec1f6e6ae

File: lab04_example/srv/KeyToCmdVel.srv (hunk 15 above)
Comment (ljden) [reply lab04_example/srv/KeyToCmdVel.srv:1]: It was given in the lab

File: lab04_example/src/velocity_command_node.cpp (hunk 14 above)
Comment (ljden) [reply lab04_example/src/velocity_command_node.cpp:29]: Added in the class

File: lab04_example/src/key_to_cmd_vel_server.cpp (hunk 12 above)
Comment (ljden) [reply lab04_example/src/key_to_cmd_vel_server.cpp:10]: it doesn't really

File: lab04_example/include/lab04_example/VelocityController.h (hunk 8 above)
Comment (ljden) [reply lab04_example/include/lab04_example/VelocityController.h:19]: 4f5e9f7b18aad97f02a4c2dc5ecaa2cec1f6e6ae

File: lab04_example/include/lab04_example/VelocityController.h (hunk 6 above)
Comment (ljden) [reply lab04_example/include/lab04_example/VelocityController.h]: 4f5e9f7b18aad97f02a4c2dc5ecaa2cec1f6e6ae

File: lab04_example/package.xml (hunk 3 above)
Comment (ljden) [reply lab04_example/package.xml]: That link is very dense

File: lab04_example/package.xml (hunk 1 above)
Comment (ljden) [reply lab04_example/package.xml]: Any benefits of MIT over BSD?

File: lab04_example/include/lab04_example/key_control.h (hunk 10 above)
Comment (ljden) [reply lab04_example/include/lab04_example/key_control.h:9]: This is code pulled from somewhere else (credited above) and waaay outside the scope of this course

File: lab04_example/src/key_input.cpp (hunk 11 above)
Comment (ljden) [reply lab04_example/src/key_input.cpp:95]: Ah, throttled log, it's very easy to find ROS commands and I don't think we need to comment things they can copy and search trivially http://wiki.ros.org/roscpp/Overview/Logging#Log_Statements

File: lab04_example/src/key_to_cmd_vel_server.cpp (hunk 12 above)
Comment (ljden) [reply lab04_example/src/key_to_cmd_vel_server.cpp:14]: 58e65d4812f3cc50574b7ac4a35970ee90ad6879ad198a97ad8a38220a446a62

File: lab04_example/src/key_to_cmd_vel_server.cpp (hunk 13 above)
Comment (ljden) [reply lab04_example/src/key_to_cmd_vel_server.cpp]: c9fd7ca22097d64d394a8c2c00a58c7ac7ab6f7b

File: lab04_example/src/key_to_cmd_vel_server.cpp (hunk 12 above)
Comment (ljden) [reply lab04_example/src/key_to_cmd_vel_server.cpp:14]: c9fd7ca22097d64d394a8c2c00a58c7ac7ab6f7b

File: lab04_example/include/lab04_example/VelocityController.h (hunk 7 above)
Comment (ljden) [reply lab04_example/include/lab04_example/VelocityController.h:36]: default value construction, probably was covered in 2500

**Comment by ljden** [comment]
I need to add a launch file + instructions to the readme

**Pull request merged**